# **Summarisation & Evasion Notebook**

# **Handover Notes:** [delete after]
- Library imports and versions are saved in environments/summarisation_evasion_env.txt
- This notebook was originally built for a macbook pro M3 chip so some settings may need to be altered depending on your machine
- All files related/ generated by this notebook can be found in notebooks/summarisation_evasion_files

### **Work progress**
1. **Complete**
- Summarise banker answers using baseline model.
- Use Local RAG pipeline to bring in relevant external documents (PRA risk definitions) to create PRA aligned summaries.
- Developed a evasion detection prototype that generates evasion scores based on bankers answers (uses baseline model, LLM- natural language inference using RoBERTa and a blended score)
- Used jpm_2025 transcripts to get the pipeline working. Validated the evasion pipeline using jpm-23-1q data (involved human labelling the answer as Direct or Evasive- file saved in notebooks/summarisation_evasion_files).

2. **Not complete**
- Visualisations e.g. how many evasive answers were there? etc - apply evasion pipeline to dataset and generate statistics on evasiveness 
- Need to test pipleine on larger data set (e.g. jpm 2023-2025) and check against HSBC to make conclusions & comment on generalisability (answering research question: How does one bank’s tone and thematic profile compare to peers? Are divergences systemic or firm specific?)
- Summarisation pipeline could be improved using a two-stage pipeline: by first extractive summarisation to capture the context and details and then a second model to reframe the summary to be PRA and evasion aligned.
- Post-processing on the output file for the PRA aligned summaries by Mistral model so they are clearer- can this output be fed into another model to extract more insights/ detect evasion or risk?
- Increase the size of the validation set for the evasion pipeline prototype (e.g. more human labelling)
- Need to fine tune the evasion pipeline to increase accuracy
- Optional extensions e.g. using Agents, more complex RAG pipeline (including more useful context for the model), validation of instances of evasion using external news sources)

# 1. **Objectives**

# **2. Set up Workspace**

In [49]:
# Import libraries
# Core python
import os
import numpy as np
import pandas as pd
import re
import json
import pathlib
from pathlib import Path
from typing import List, Dict, Any 
import csv
import math

# NLP & Summarisation
from transformers import pipeline, AutoTokenizer, AutoModelForSequenceClassification
from llama_cpp import Llama 
import torch
import torch.nn.functional as F

# Retrieval
from sentence_transformers import SentenceTransformer 

# ML
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

# Visualisations
import matplotlib.pyplot as plt
import seaborn as sns 

# Set SEED.
SEED = 42


# **3. Load the dataset**

In [50]:
# Load the dataset.
jpm_2025_df = pd.read_csv('../data/processed/jpm/all_jpm_2025.csv')

# View the data.
jpm_2025_df.head()

question_number  answer_number   speaker_name  \
0                1            NaN      Ken Usdin   
1                1            1.0  Jeremy Barnum   
2                2            NaN      Ken Usdin   
3                2            1.0  Jeremy Barnum   
4                2            2.0    Jamie Dimon   

                                 role              company  \
0                             analyst  Autonomous Research   
1             Chief Financial Officer        JPMorganChase   
2                             analyst  Autonomous Research   
3             Chief Financial Officer        JPMorganChase   
4  Chairman & Chief Executive Officer        JPMorganChase   

                                             content  year quarter  \
0  Good morning, Jeremy. Wondering if you could s...  2025      Q1   
1  Sure, Ken. So I mean, at a high level, I would...  2025      Q1   
2  Yeah. And just one question on the NII ex. Mar...  2025      Q1   
3  Yeah, that's a good question, Ken. You're righ...  2025      Q1   
4                            In the curve basically.  2025      Q1   

                                          source_pdf  
0  data/raw/jpm/jpm-1q25-earnings-call-transcript...  
1  data/raw/jpm/jpm-1q25-earnings-call-transcript...  
2  data/raw/jpm/jpm-1q25-earnings-call-transcript...  
3  data/raw/jpm/jpm-1q25-earnings-call-transcript...  
4  data/raw/jpm/jpm-1q25-earnings-call-transcript...

# **4. Preprocessing**

- Used all_jpm_2025.csv dataset
- Preliminary preprocessing to label roles as analyst vs banker (invalid roles were corrected) to make downstream analysis easier. Created a new column 'role_normalised'.

In [51]:
# View speaker roles.
jpm_2025_df['role'].unique()

array(['analyst', 'Chief Financial Officer',
       'Chairman & Chief Executive Officer',
       'And then some. Theres a lot of value added.', 'Okay'],
      dtype=object)

In [52]:
# View rows with invalid roles.
valid_roles = 'analyst', 'Chief Financial Officer', 'Chairman & Chief Executive Officer'
invalid_roles_df = jpm_2025_df[~jpm_2025_df['role'].isin(valid_roles)]

# Number of rows with invalid roles.
print('Number of rows:', invalid_roles_df.shape[0])

# View the rows.
invalid_roles_df.head()

Number of rows: 2


question_number  answer_number                            speaker_name  \
201               35            5.0  Chief Financial Officer, JPMorganChase   
205               36            3.0  Chief Financial Officer, JPMorganChase   

                                            role             company  \
201  And then some. Theres a lot of value added.       JPMorganChase   
205                                         Okay  there you have it.   

                                               content  year quarter  \
201  Yeah. And obviously, I mean, we're not going t...  2025      Q2   
205  But it's not like I thought it would do badly,...  2025      Q2   

                                            source_pdf  
201  data/raw/jpm/jpm-2q25-earnings-call-transcript...  
205  data/raw/jpm/jpm-2q25-earnings-call-transcript...

In [53]:
# Input the correct role information.
jpm_2025_df.at[205, 'role'] = 'Chief Financial Officer'
jpm_2025_df.at[209, 'role'] = 'Chief Financial Officer'

# Verify the roles have been updated.
jpm_2025_df['role'].unique()

array(['analyst', 'Chief Financial Officer',
       'Chairman & Chief Executive Officer',
       'And then some. Theres a lot of value added.'], dtype=object)

In [54]:
# Define role mapping.
role_map = {
    'analyst': 'analyst',
    'Chief Financial Officer': 'banker',
    'Chairman & Chief Executive Officer': 'banker'
}

# Apply to dataset.
jpm_2025_df['role_normalised'] = jpm_2025_df['role'].map(role_map)

In [55]:
# View the dataset.
jpm_2025_df.head()

question_number  answer_number   speaker_name  \
0                1            NaN      Ken Usdin   
1                1            1.0  Jeremy Barnum   
2                2            NaN      Ken Usdin   
3                2            1.0  Jeremy Barnum   
4                2            2.0    Jamie Dimon   

                                 role              company  \
0                             analyst  Autonomous Research   
1             Chief Financial Officer        JPMorganChase   
2                             analyst  Autonomous Research   
3             Chief Financial Officer        JPMorganChase   
4  Chairman & Chief Executive Officer        JPMorganChase   

                                             content  year quarter  \
0  Good morning, Jeremy. Wondering if you could s...  2025      Q1   
1  Sure, Ken. So I mean, at a high level, I would...  2025      Q1   
2  Yeah. And just one question on the NII ex. Mar...  2025      Q1   
3  Yeah, that's a good question, Ken. You're righ...  2025      Q1   
4                            In the curve basically.  2025      Q1   

                                          source_pdf role_normalised  
0  data/raw/jpm/jpm-1q25-earnings-call-transcript...         analyst  
1  data/raw/jpm/jpm-1q25-earnings-call-transcript...          banker  
2  data/raw/jpm/jpm-1q25-earnings-call-transcript...         analyst  
3  data/raw/jpm/jpm-1q25-earnings-call-transcript...          banker  
4  data/raw/jpm/jpm-1q25-earnings-call-transcript...          banker

# **5. Summarisation**

## **5.1 Baseline**

- Initial model exploration using BART and mistral-7B-instruct to summarise banker's answers (no additional context given to model)

### **5.1.1 BART**

In [56]:
# Filter data to banker answers only.
banker_answers = jpm_2025_df[jpm_2025_df['role_normalised'] == 'banker']['content'].tolist()
print(banker_answers[0][:200])

Sure, Ken. So I mean, at a high level, I would say that obviously, some of the salient news flow is quite recent. So, we've done some soundings and some checking both on the consumer side and on the w


In [57]:
# Summarisation baseline (BART)
bart = pipeline('summarization', model='facebook/bart-large-cnn')

sample_text = banker_answers[0]
summary_bart = bart(sample_text, max_length=80, min_length=30, do_sample=False)
print('Original:', sample_text[:400])
print('Summary:', summary_bart[0]['summary_text'])

Device set to use mps:0


Original: Sure, Ken. So I mean, at a high level, I would say that obviously, some of the salient news flow is quite recent. So, we've done some soundings and some checking both on the consumer side and on the wholesale side. I think on the consumer side, the thing to check is the spending data. And to be honest, the main thing that we see there, what would appear to be a certain amount of frontloading of sp
Summary: The main thing that we see there, what would appear to be a certain amount of frontloading of spending ahead of people expecting price increases from tariffs. So ironically, that's actually somewhat supportive, all else equal. In terms of our corporate clients, obviously, they've been reacting to the changes in tariff policy.


- bart was able to extract ket ideas, focussing on fronloading of spending and tariff policy. 
- Compressed the response into two sentences and the summary is coherent, removing filler phrases.
- However, the summary is not fully neutral (e.g. includes ironically) and preserves tone
- Also there is a loss of context- e.g. consumer side vs wholesale side distinction is no longer explicit.

In [58]:
# Prompt conditioning to make PRA relevant.
prompt = "Summarise this answer, focusing on risk, capital and evasion of detail: " + sample_text
summary_bart_prompted = bart(prompt, max_length=80, min_length=30)
print('Original:', sample_text[:400])
print('Summary:', summary_bart_prompted[0]['summary_text'])

Original: Sure, Ken. So I mean, at a high level, I would say that obviously, some of the salient news flow is quite recent. So, we've done some soundings and some checking both on the consumer side and on the wholesale side. I think on the consumer side, the thing to check is the spending data. And to be honest, the main thing that we see there, what would appear to be a certain amount of frontloading of sp
Summary: Corporates are taking a wait-and-see approach to tariff policy. Some sectors are going to be much more exposed than others. Small business and smaller corporates are probably a little more challenged.


- Prompted summary shifts emphasis and includes interpretation around risk, even though those words were no explicit in the original
- This version is more aligned to evasion detection but moves away from concrete detail 
- Improved approach would be to have a two stage-pipeline: first extractive summarisation to capture the context and details and then a second model to reframe the summary to be PRA and evasion aligned.

### **5.1.2 Mistral-7B-Instruct**

- Mistral model: mistral-7b-instruct-v0.1.Q4_K_M.gguf
- Mistral-7B-Instruct model download: https://huggingface.co/TheBloke/Mistral-7B-Instruct-v0.1-GGUF?show_file_info=mistral-7b-instruct-v0.1.Q4_K_M.gguf
- Also saved in shared team folder models

In [59]:
# Summarisation baseline (Mistral-7B-Instruct) with basic prompt.
llm = Llama(model_path='/Users/laurenbrixey/Documents/Data Science Career Accelerator/Project Submissions/Course 3/topic_project_4.1/mistral-7b-instruct-v0.1.Q4_K_M.gguf',
            n_ctx=4096, n_gpu_layers=-1, verbose=False, seed=SEED)  # change path as needed 

prompt = f"<s>[INST] Summarise the following answer in 2 sentences, focusing on concrete facts. Avoid opinions.\n\n{sample_text}\n[/INST]"

output = llm.create_chat_completion(
    messages=[{'role': 'user', 'content': prompt}],
    max_tokens=180,
    temperature=0.1,
    stop=['</s>']
)

summary_mistral = output['choices'][0]['message']['content'].strip()  

print('Original:', sample_text[:400])
print('Summary:', summary_mistral)

llama_context: n_ctx_per_seq (4096) < n_ctx_train (32768) -- the full capacity of the model will not be utilized
ggml_metal_init: skipping kernel_get_rows_bf16                     (not supported)
ggml_metal_init: skipping kernel_set_rows_bf16                     (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32                   (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32_c4                (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32_1row              (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32_l4                (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_bf16                  (not supported)
ggml_metal_init: skipping kernel_mul_mv_id_bf16_f32                (not supported)
ggml_metal_init: skipping kernel_mul_mm_bf16_f32                   (not supported)
ggml_metal_init: skipping kernel_mul_mm_id_bf16_f16                (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h64 

Original: Sure, Ken. So I mean, at a high level, I would say that obviously, some of the salient news flow is quite recent. So, we've done some soundings and some checking both on the consumer side and on the wholesale side. I think on the consumer side, the thing to check is the spending data. And to be honest, the main thing that we see there, what would appear to be a certain amount of frontloading of sp
Summary: The speaker is discussing the impact of recent news flow on the consumer and corporate sides. On the consumer side, there has been some frontloading of spending ahead of expected price increases from tariffs, which may distort the data and make it difficult to draw larger conclusions. On the corporate side, clients are reacting to changes in tariff policy by shifting their focus towards short-term work and optimizing supply chains. The speaker characterizes the attitude of corporate clients as a wait-and-see attitude, with smaller clients and smaller corporates being more c

- Preserves details and nuance and is more contextual and interpretive than the BART baseline model.
- However, the result is longer with heavier phrasing and includes phrases like 'distort the data' which is not explicit in the original.

In [60]:
# Summarisation baseline (Mistral-7B-Instruct) with more detailed prompt.
llm = Llama(model_path='/Users/laurenbrixey/Documents/Data Science Career Accelerator/Project Submissions/Course 3/topic_project_4.1/mistral-7b-instruct-v0.1.Q4_K_M.gguf',
            n_ctx=4096, n_gpu_layers=-1, verbose=False, seed=SEED)  # change path as needed 

prompt = f"<s>[INST] Summarise the following answer in 2 sentences, focusing on concrete facts. Avoid opinions. Focus on risk, capital and evasion of detail.\n\n{sample_text}\n[/INST]"

output = llm.create_chat_completion(
    messages=[{'role': 'user', 'content': prompt}],
    max_tokens=180,
    temperature=0.1,
    stop=['</s>']
)

summary_mistral_prompted = output['choices'][0]['message']['content'].strip()  

print('Original:', sample_text[:400])
print('Summary:', summary_mistral_prompted)

llama_context: n_ctx_per_seq (4096) < n_ctx_train (32768) -- the full capacity of the model will not be utilized
ggml_metal_init: skipping kernel_get_rows_bf16                     (not supported)
ggml_metal_init: skipping kernel_set_rows_bf16                     (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32                   (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32_c4                (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32_1row              (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32_l4                (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_bf16                  (not supported)
ggml_metal_init: skipping kernel_mul_mv_id_bf16_f32                (not supported)
ggml_metal_init: skipping kernel_mul_mm_bf16_f32                   (not supported)
ggml_metal_init: skipping kernel_mul_mm_id_bf16_f16                (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h64 

Original: Sure, Ken. So I mean, at a high level, I would say that obviously, some of the salient news flow is quite recent. So, we've done some soundings and some checking both on the consumer side and on the wholesale side. I think on the consumer side, the thing to check is the spending data. And to be honest, the main thing that we see there, what would appear to be a certain amount of frontloading of sp
Summary: The speaker is discussing the impact of recent news flow on the consumer and corporate sides of their business. On the consumer side, they have observed some frontloading of spending ahead of expected price increases from tariffs, which may distort data and make it difficult to draw larger conclusions. On the corporate side, clients are shifting their focus towards optimizing supply chains and responding to the current environment, rather than prioritizing more strategic work. The speaker notes that smaller clients and smaller corporates may be more challenged than larger o

- This summary brings in risk- language and is closer to the task objective.
- However, some interpretations are generated by the model rather than explicitly detailed in the answer

## **5.2 Adding Context**

Retrieve PRA risk categories to give greater PRA focus to summaries (local RAG loop).
- measure cosine similarity between transcript chunks and PRA risk categories (vectors)
- retrieve the top 2-3 most relevant risk categories 
- prepend them to the summarisation prompt to make summaries PRA-aligned instead of just summarised answers

- Attempting to use BART resulted in prompt echoing.
- New attempt using Mistral-7B-Instruct.
- Using sentence-BERT vs TF-IDF for vectorisation.

### **5.2.1 Mistral-7B-Instruct**

**Process**
- Performed some light cleaning of the transcript to remove whitespace.
- Split the transcript into smaller chunks that the model can summarise to avoid truncation
- Loaded the PRA categories csv file (contains category and definition)
- Embedded the PRA categories and chunks, evaluated the similarity to extract the PRA risk categories that were relevant to the text
- Summarised the chunk using detailed prompted and relevant PRA categories as additional context. 

**Output File**:
- The output file of this can be found in notebooks/summarisation_evasion_files, name = jpm_mistral_pra_summary.json
- It is in the format: summary, evidence, PRA category that relates to summary and reasoning for selecting these categories.

- Needed to use a lot of fine tuning for the prompt and set strict rules for the model
- Need to be very clear about the output expected or else the model deviates a lot, especially as it processes more data.
- Include lines about lack of evidence if not the model may hallucinate

In [61]:
# Function to remove whitespace in text.
def clean_text(text: str):
    return re.sub(r'\s+', ' ', text).strip()

In [62]:
# Function to split the transcript into smaller chunks.
def chunk_text(text: str, max_chars: int = 6000):
    sentences = re.split(r'(?<=[.!?])\s+', text.strip()) # split into sentences 
    chunks, current_chunk, current_len = [], [], 0 # list of chunks, sentences collecting for current chunk, character count for current chunk

    for s in sentences:
        if current_len + len(s) + 1 <= max_chars: # if the characters of current chunk + new sentence is below the limit:
            current_chunk.append(s) # add sentence to current chunk 
            current_len += len(s) + 1 # update running character count 
        
        else: # if the characters is above the limit:
            chunks.append(' '.join(current_chunk)) # add the current chunk to the final chunk list
            current_chunk, current_len = [s], len(s) # start a new chunk containing the sentence and update current len

    if current_chunk:
        chunks.append(' '.join(current_chunk)) # add any sentences in current chunk after loop ends 

    return chunks 

In [63]:
# Function to load PRA categories and definitions from CSV.
def load_pra_categories(path: Path):
    with open(path, newline='', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        return [
            (row.get('category', '').strip(), [row.get('definition', '').strip()])
            for row in reader if row.get('category')
        ]

In [64]:
# Build a Sentence-BERT embedding index for PRA categories.
def build_embedding_index(pra_categories):
    embedder = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
    docs = [f"{name} {' '.join(defs)}" for name, defs in pra_categories]
    pra_risk_embeddings = embedder.encode(docs, batch_size=32, normalize_embeddings=True)

    return embedder, np.asarray(pra_risk_embeddings)

In [65]:
# Function to find the relevant PRA categories to the transcript chunks.
def find_rel_categories(chunk, pra_categories, embedder, pra_risk_embeddings, top_k=2):
    query_vec = embedder.encode([chunk], normalize_embeddings=True) # turns chunk into embedding
    sims = cosine_similarity(query_vec, pra_risk_embeddings).ravel() # compares the chunk to each category doc 
    top_indices = np.argsort(-sims)[:top_k] # sorts scores descending and selected top k cateogories 

    return [pra_categories[i] for i in top_indices]

In [66]:
# Function to parse JSON
def parse_tagged_json(raw):
    m = re.search(r"<json>\s*(\{[\s\S]*?\})\s*</json>", raw, flags=re.IGNORECASE)
    if not m:
        return None
    try:
        return json.loads(m.group(1))
    except json.JSONDecodeError:
        return None

In [67]:
# Function to summarise the text chunks.
def summarise_chunk(model, chunk, relevant_categories, max_evidence=5):

    # Build PRA notes (limit to 2 bullets per category)
    lines = []
    for name, definition in relevant_categories:
        lines.append(f'- {name}:')
        for d in list(definition)[:2]:
            lines.append(f'- {d}')
    notes_block = '\n'.join(lines)

    system_prompt = (
        "You are a careful data extraction model. "
        "Return ONLY valid JSON wrapped in <json>...</json> tags."
    )

    user_prompt = f"""
TRANSCRIPT:
{chunk}

PRA NOTES:
{notes_block}

TASK:
Return JSON ONLY, wrapped exactly like this:
<json>{{"summary": "...", "evidence": ["..."], "pra_categories": [{{"category":"...","why":"..."}}]}}</json>

RULES:
- 4-6 sentence neutral summary.
- Up to {max_evidence} evidence bullets (quotes/facts).
- 1-3 pra_categories objects.
- If evidence is lacking, use a single bullet "Insufficient evidence".
- Only choose categories supported by the evidence.
""".strip()

    response = model.create_chat_completion(
        messages=[
            {'role': 'system', 'content': system_prompt},
            {'role': 'user', 'content': user_prompt},
        ],
        temperature=0.2,
        top_p=0.9,
        max_tokens=700,
        repeat_penalty=1.1,
    )

    raw = (response['choices'][0]['message']['content'] or '').strip()

    # Parse the tagged JSON
    parsed = parse_tagged_json(raw)

    # Fallback if model didn’t follow instructions
    if not parsed:
        return (
            {'summary': '', 'evidence': ['Insufficient evidence'], 'pra_categories': []},
            raw,
        )

    # Light coercion to guarantee keys exist
    result = {
        'summary': parsed.get('summary', '') or '',
        'evidence': parsed.get('evidence', []) or [],
        'pra_categories': parsed.get('pra_categories', []) or []
    }
    return result, raw

In [68]:
# Define variables.
MODEL_PATH = '/Users/laurenbrixey/Documents/Data Science Career Accelerator/Project Submissions/Course 3/topic_project_4.1/mistral-7b-instruct-v0.1.Q4_K_M.gguf'
PRA_NOTES_PATH = '../data/RAG-resources/PRA_risk_categories.csv'
TRANSCRIPT_PATH = '../data/processed/jpm/all_jpm_2025.csv'
OUTPUT_PATH = pathlib.Path('../notebooks/summarisation_evasion_files/jpm_mistral_pra_summary_raw.json')
TOP_K = 2

In [69]:
# Runner code.
pra_categories = load_pra_categories(Path(PRA_NOTES_PATH))
embedder, category_embeddings = build_embedding_index(pra_categories)

# Load and chunk transcript
transcript_text = Path(TRANSCRIPT_PATH).read_text(encoding='utf-8')
transcript_chunks = chunk_text(transcript_text)

n_threads = max(4, (os.cpu_count() or 8) - 2)

# Define the model.
model = Llama(
    model_path=str(MODEL_PATH),
    n_ctx=4096,
    n_gpu_layers=20,
    chat_format='mistral-instruct',
    n_threads=n_threads,
)

raw_outputs = []

for i, chunk in enumerate(transcript_chunks, 1):
    try:
        top_categories = find_rel_categories(
            chunk, pra_categories, embedder, category_embeddings, top_k=TOP_K
        )
        _, raw = summarise_chunk(
            model, chunk, top_categories, max_evidence=5
        )
        raw_outputs.append({'chunk': i, 'raw': raw})

    except Exception:
        raw_outputs.append({'chunk': i, 'raw': ''})

final_output = {'raw_outputs': raw_outputs}

OUTPUT_PATH.write_text(json.dumps(final_output, indent=2, ensure_ascii=False), encoding='utf-8')
print(f'Wrote final JSON to: {OUTPUT_PATH.resolve()}')

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
llama_model_load_from_file_impl: using device Metal (Apple M3) - 3559 MiB free
llama_model_loader: loaded meta data with 20 key-value pairs and 291 tensors from /Users/laurenbrixey/Documents/Data Science Career Accelerator/Project Submissions/Course 3/topic_project_4.1/mistral-7b-instruct-v0.1.Q4_K_M.gguf (version GGUF V2)
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = mistralai_mistral-7b-instruct-v0.1
llama_model_loader: - kv   2

Wrote final JSON to: /Users/laurenbrixey/Documents/GitHub Repositories/cam_ds_ep_FinSight/notebooks/summarisation_evasion_files/jpm_mistral_pra_summary_raw.json


- Need to preprocess the output so it is visually clearer (summary, evidence, PRA categories (name & why the model chose this))
- Can this information be fed to the model again and can it detect any early PRA risk indicators?

# **6. Evasion Detection Pipeline**

1. **Baseline Evasion score** (rule-based) is made up of three components:
- **Cosine similarity**- similarity of the question and answer, lower similarity = more evasive
- **Numeric specificity check**- does the question require a number, if so does the answer contain a number?, e.g. requests for financial data
- **Evasive phrases**- does the answer contain evasive phrases?, presence = more evasive

2. **LLM evasion score** (RoBERTa-MNLI) uses entailment/neutral/contradiction between the question and answer
- Lower entailment (and higher neutral + contradiction) = more evasive
  
3. **Blended evasion score** combines both scores including a weight for the LLM component
- Rationale is that baseline enforces precision while the LLM will capture semantics

### **Data Preprocessing**

In [221]:
# Load dataset
all_jpm_2023_2025 = pd.read_csv('../data/processed/jpm/all_jpm_2023_2025.csv')

# View dataset.
display(all_jpm_2023_2025.head())

# Number of rows.
print('Number of rows:', all_jpm_2023_2025.shape[0])

section  question_number  answer_number   speaker_name  \
0  presentation              NaN            NaN  Jeremy Barnum   
1            qa              NaN            NaN  Steven Chubak   
2            qa              NaN            NaN  Jeremy Barnum   
3            qa              1.0            NaN  Steven Chubak   
4            qa              1.0            1.0    Jamie Dimon   

                                 role               company  \
0             Chief Financial Officer         JPMorganChase   
1                             analyst    Wolfe Research LLC   
2             Chief Financial Officer  JPMorgan Chase & Co.   
3                             analyst    Wolfe Research LLC   
4  Chairman & Chief Executive Officer  JPMorgan Chase & Co.   

                                             content  year quarter  \
0  Thanks, and good morning, everyone. The presen...  2023      Q1   
1                                 Hey, good morning.  2023      Q1   
2                               Good morning, Steve.  2023      Q1   
3  So, Jamie, I was actually hoping to get your p...  2023      Q1   
4  Well, I think you were already kind of complet...  2023      Q1   

   is_pleasantry                                         source_pdf  
0          False  data/raw/jpm/.ipynb_checkpoints/jpm-1q23-earni...  
1           True  data/raw/jpm/.ipynb_checkpoints/jpm-1q23-earni...  
2           True  data/raw/jpm/.ipynb_checkpoints/jpm-1q23-earni...  
3          False  data/raw/jpm/.ipynb_checkpoints/jpm-1q23-earni...  
4          False  data/raw/jpm/.ipynb_checkpoints/jpm-1q23-earni...

Number of rows: 1411


In [222]:
# Remove pleasantries.
all_jpm_2023_2025 = all_jpm_2023_2025[all_jpm_2023_2025['is_pleasantry'] == False]
print('Number of rows:', all_jpm_2023_2025.shape[0])

Number of rows: 1241


In [223]:
# Check content column.
print('Number of rows with no content:', all_jpm_2023_2025['content'].isna().sum())

Number of rows with no content: 23


In [225]:
# Drop rows with no content.
all_jpm_2023_2025 = all_jpm_2023_2025.dropna(subset=['content'])

In [226]:
# Check content column.
print('Number of rows with no content:', all_jpm_2023_2025['content'].isna().sum())

Number of rows with no content: 0


In [227]:
# View roles.
all_jpm_2023_2025['role'].unique()

array(['Chief Financial Officer', 'analyst',
       'Chairman & Chief Executive Officer',
       'And then some. Theres a lot of value added.', 'Okay',
       "We're fundamentally", 'Thanks', 'Almost no chance.'], dtype=object)

- Some text has leaked into role column.

In [228]:
# View rows with invalid roles. 
valid_roles = 'analyst', 'Chief Financial Officer', 'Chairman & Chief Executive Officer'
invalid_roles_df = all_jpm_2023_2025[~all_jpm_2023_2025['role'].isin(valid_roles)]
invalid_roles_df.head(10)

section  question_number  answer_number  \
305       qa             22.0            4.0   
309       qa             23.0            3.0   
650       qa             10.0            3.0   
924       qa              8.0            2.0   
1059      qa             22.0            4.0   
1063      qa             23.0            3.0   
1274      qa             23.0            1.0   

                                           speaker_name  \
305              Chief Financial Officer, JPMorganChase   
309              Chief Financial Officer, JPMorganChase   
650   Who knows how important politics are in all th...   
924       Chief Financial Officer, JPMorgan Chase & Co.   
1059             Chief Financial Officer, JPMorganChase   
1063             Chief Financial Officer, JPMorganChase   
1274  Chairman & Chief Executive Officer, JPMorgan C...   

                                             role  \
305   And then some. Theres a lot of value added.   
309                                          Okay   
650                           We're fundamentally   
924                                        Thanks   
1059  And then some. Theres a lot of value added.   
1063                                         Okay   
1274                            Almost no chance.   

                                                company  \
305                                       JPMorganChase   
309                                  there you have it.   
650   as I said, I think on the press call, happy to...   
924                                              Glenn.   
1059                                      JPMorganChase   
1063                                 there you have it.   
1274                                      JPMorganChase   

                                                content  year quarter  \
305   Yeah. And obviously, I mean, we're not going t...  2025      Q2   
309   But it's not like I thought it would do badly,...  2025      Q2   
650   little bit cautious about the pull-forward dyn...  2024      Q1   
924   Operator: Next, we'll go to the line of Matt O...  2024      Q2   
1059  Yeah. And obviously, I mean, we're not going t...  2025      Q2   
1063  But it's not like I thought it would do badly,...  2025      Q2   
1274  Well, but having – it's very important. While ...  2024      Q3   

      is_pleasantry                                         source_pdf  
305           False  data/raw/jpm/.ipynb_checkpoints/jpm-2q25-earni...  
309           False  data/raw/jpm/.ipynb_checkpoints/jpm-2q25-earni...  
650           False  data/raw/jpm/jpm-1q24-earnings-call-transcript...  
924           False  data/raw/jpm/jpm-2q24-earnings-call-transcript...  
1059          False  data/raw/jpm/jpm-2q25-earnings-call-transcript...  
1063          False  data/raw/jpm/jpm-2q25-earnings-call-transcript...  
1274          False  data/raw/jpm/jpm-3q24-earnings-conference-call...

In [229]:
# Input the correct role information.
all_jpm_2023_2025.loc[[305, 309, 924, 1059, 1063], 'role'] = 'Chief Financial Officer'
all_jpm_2023_2025.loc[[1274], 'role'] = 'Chairman & Chief Executive Officer'

# Drop nonsence row.
all_jpm_2023_2025 = all_jpm_2023_2025.drop(index=650)

In [230]:
# Check the roles have been updated.
all_jpm_2023_2025['role'].unique()

array(['Chief Financial Officer', 'analyst',
       'Chairman & Chief Executive Officer'], dtype=object)

In [231]:
# Normalise role names.
role_map = {
    'analyst': 'analyst',
    'Chief Financial Officer': 'banker',
    'Chairman & Chief Executive Officer': 'banker'
}

# Map roles.
all_jpm_2023_2025['role_normalised'] = all_jpm_2023_2025['role'].map(role_map)

In [232]:
# View dataset.
display(all_jpm_2023_2025.head())
print('Number of rows:', all_jpm_2023_2025.shape[0])

section  question_number  answer_number   speaker_name  \
0  presentation              NaN            NaN  Jeremy Barnum   
3            qa              1.0            NaN  Steven Chubak   
4            qa              1.0            1.0    Jamie Dimon   
5            qa              1.0            1.0  Steven Chubak   
6            qa              1.0            2.0    Jamie Dimon   

                                 role               company  \
0             Chief Financial Officer         JPMorganChase   
3                             analyst    Wolfe Research LLC   
4  Chairman & Chief Executive Officer  JPMorgan Chase & Co.   
5                             analyst    Wolfe Research LLC   
6  Chairman & Chief Executive Officer  JPMorgan Chase & Co.   

                                             content  year quarter  \
0  Thanks, and good morning, everyone. The presen...  2023      Q1   
3  So, Jamie, I was actually hoping to get your p...  2023      Q1   
4  Well, I think you were already kind of complet...  2023      Q1   
5  Got it. And just in terms of appetite for the ...  2023      Q1   
6                                          Oh, yeah.  2023      Q1   

   is_pleasantry                                         source_pdf  \
0          False  data/raw/jpm/.ipynb_checkpoints/jpm-1q23-earni...   
3          False  data/raw/jpm/.ipynb_checkpoints/jpm-1q23-earni...   
4          False  data/raw/jpm/.ipynb_checkpoints/jpm-1q23-earni...   
5          False  data/raw/jpm/.ipynb_checkpoints/jpm-1q23-earni...   
6          False  data/raw/jpm/.ipynb_checkpoints/jpm-1q23-earni...   

  role_normalised  
0          banker  
3         analyst  
4          banker  
5         analyst  
6          banker

Number of rows: 1217


In [239]:
# Split into test, validation and training.
jpm_test = all_jpm_2023_2025[(all_jpm_2023_2025['year'] == 2023) & (all_jpm_2023_2025['quarter'].isin(['Q1', 'Q2']))]
jpm_val = all_jpm_2023_2025[(all_jpm_2023_2025['year'] == 2023) & (all_jpm_2023_2025['quarter'].isin(['Q3', 'Q4']))]
jpm_predict = all_jpm_2023_2025[(all_jpm_2023_2025['year'].isin([2024, 2025]))]

print('Number of test rows:', jpm_test.shape[0])
print('Number of validation rows:', jpm_val.shape[0])
print('Number of prediction rows:', jpm_train.shape[0])

Number of test rows: 262
Number of validation rows: 202
Number of prediction rows: 753


In [240]:
# Pair each analyst question with all the banker's answers (one row per banker answer).
def create_qa_pairs(df):
    
    # Keep only the Q&A section.
    qa_df = df[df['section'].astype(str).str.lower() == 'qa'].copy()

    # Split into roles.
    analyst_rows = qa_df[qa_df['role_normalised'] == 'analyst']
    banker_rows = qa_df[qa_df['role_normalised'] == 'banker']

    # Build full question text per question_number (concatenate all analyst utterances)
    question_text_map = (
        analyst_rows.groupby('question_number')['content']
        .apply(lambda parts: ' '.join(parts.astype(str)))
        .rename('question')
    )

    # Ensure bankers have an answer_number — if missing, assign sequentially per question
    if 'answer_number' not in banker_rows.columns or banker_rows['answer_number'].isna().any():
        banker_rows = banker_rows.sort_index().copy()
        banker_rows['answer_number'] = banker_rows.groupby('question_number').cumcount() + 1

    # Combine multiple banker utterances belonging to the same answer
    banker_answers = (
        banker_rows.groupby(['question_number', 'answer_number'])
        .agg({
            'content': lambda parts: ' '.join(parts.astype(str)),
            'speaker_name': 'first',
            'role': 'first',
            'role_normalised': 'first',
            'year': 'first',
            'quarter': 'first',
            'source_pdf': 'first'
        })
        .rename(columns={'content': 'answer'})
        .reset_index()
    )

    # Merge the question text back onto each answer row
    qa_pairs = banker_answers.merge(
        question_text_map,
        left_on='question_number',
        right_index=True,
        how='left'
    )

    # Order columns for readability
    column_order = [
        'question_number', 'answer_number', 'question', 'answer',
        'speaker_name', 'role', 'role_normalised',
        'year', 'quarter', 'source_pdf'
    ]
    qa_pairs = qa_pairs.reindex(columns=[c for c in column_order if c in qa_pairs.columns])

    # Sort nicely and reset index
    qa_pairs = qa_pairs.sort_values(['question_number', 'answer_number']).reset_index(drop=True)

    return qa_pairs 

In [241]:
# Create q&A pairs.
jpm_test_qa = create_qa_pairs(jpm_test)
jpm_val_qa = create_qa_pairs(jpm_val)
jpm_predict_qa = create_qa_pairs(jpm_predict)

In [242]:
# Remove duplicated rows.
jpm_predict_qa = jpm_predict_qa.drop_duplicates(subset=['answer'])

In [244]:
# View number of examples.
print('Number of test examples:', jpm_test_qa.shape[0])
print('Number of validation examples:', jpm_val_qa.shape[0])
print('Number of prediction examples:', jpm_predict_qa.shape[0])

Number of test examples: 57
Number of validation examples: 49
Number of prediction examples: 279


In [245]:
# Save the cleaned datasets.
jpm_test_qa.to_csv('../data/processed/jpm/cleaned/jpm_test_qa.csv')
jpm_val_qa.to_csv('../data/processed/jpm/cleaned/jpm_val_qa.csv')
jpm_predict_qa.to_csv('../data/processed/jpm/cleaned/jpm_predict_qa.csv')        

- Human label the validation and test dataset with evasive or direct labels.

### **Load the Clean & Labelled Datasets**

In [ ]:
# Load the training and labelled validation & test datasets.
jpm_predict_qa = pd.read_csv('../data/processed/jpm/cleaned/jpm_predict_qa.csv')
jpm_test_qa_labelled = pd.read_csv('../data/processed/jpm/cleaned/jpm_test_qa_labelled.csv')
jpm_val_qa_labelled = pd.read_csv('../data/processed/jpm/cleaned/jpm_val_qa_labelled.csv')

### **LLM Model Set-up**

In [74]:
# Model name checkpoints.
roberta_name = 'roberta-large-mnli'
deberta_name = 'microsoft/deberta-large-mnli'
zs_deberta_name = 'MoritzLaurer/deberta-v3-large-zeroshot-v2.0'

# Load tokenizers and models.
roberta_tokenizer = AutoTokenizer.from_pretrained(roberta_name)
roberta = AutoModelForSequenceClassification.from_pretrained(roberta_name)

deberta_tokenizer = AutoTokenizer.from_pretrained(deberta_name)
deberta = AutoModelForSequenceClassification.from_pretrained(deberta_name)

zs_deberta_tokenizer = AutoTokenizer.from_pretrained(zs_deberta_name)
zs_deberta = AutoModelForSequenceClassification.from_pretrained(zs_deberta_name)

Some weights of the model checkpoint at roberta-large-mnli were not used when initializing RobertaForSequenceClassification: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of the model checkpoint at microsoft/deberta-large-mnli were not used when initializing DebertaForSequenceClassification: ['config']
- This IS expected if you are initializing DebertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. ini

In [75]:
# Verify label order per model.
print("roberta id2label:", roberta.config.id2label)
print("deberta id2label:", deberta.config.id2label)
print("zs_deberta id2label:", zs_deberta.config.id2label)

roberta id2label: {0: 'CONTRADICTION', 1: 'NEUTRAL', 2: 'ENTAILMENT'}
deberta id2label: {0: 'CONTRADICTION', 1: 'NEUTRAL', 2: 'ENTAILMENT'}
zs_deberta id2label: {0: 'entailment', 1: 'not_entailment'}


- Roberta and deberta have the standard 3 MNLI labels whereas zero shot deberta is binary.

In [76]:
# Add models and tokenizers to dictionary.
models_and_tokenizers = {
        'roberta': (roberta, roberta_tokenizer),
        'deberta': (deberta, deberta_tokenizer),
        'zs_deberta': (zs_deberta, zs_deberta_tokenizer)
        }

In [77]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")  # no MPS
for model, tok in models_and_tokenizers.values():
    model.to(device).eval()

ggml_metal_free: deallocating
ggml_metal_mem_pool_free: freeing memory pool, num heaps = 0 (total = 0)
ggml_metal_mem_pool_free: freeing memory pool, num heaps = 0 (total = 0)
ggml_metal_mem_pool_free: freeing memory pool, num heaps = 0 (total = 0)
ggml_metal_mem_pool_free: freeing memory pool, num heaps = 0 (total = 0)
ggml_metal_mem_pool_free: freeing memory pool, num heaps = 0 (total = 0)
ggml_metal_mem_pool_free: freeing memory pool, num heaps = 0 (total = 0)
ggml_metal_mem_pool_free: freeing memory pool, num heaps = 0 (total = 0)
ggml_metal_mem_pool_free: freeing memory pool, num heaps = 0 (total = 0)


### **Baseline Evasion Score Functions**

In [78]:
# List of evasive phrases
EVASIVE_PHRASES = [
    r"\btoo early\b",
    r"\bcan't (?:comment|share|discuss)\b",
    r"\bwon't (?:comment|share|provide)\b",
    r"\bno (?:update|comment)\b",
    r"\bwe (?:don't|do not) (?:break out|provide guidance)\b",
    r"\bnot (?:going to|able to) (?:comment|share|provide)\b",
    r"\bwe'll (?:come back|circle back)\b",
    r"\bnot something we disclose\b",
    r"\bas (?:we|I) (?:said|mentioned)\b",
    r"\bgenerally speaking\b",
    r"\bit's premature\b",
    r"\bit's difficult to say\b",
    r"\bI (?:wouldn't|won't) want to (?:speculate|get into)\b",
    r"\bI (?:think|guess|suppose)\b",
    r"\bkind of\b",
    r"\bsort of\b",
    r"\baround\b",
    r"\broughly\b",
    r"\bwe (?:prefer|plan) not to\b",
    r"\bwe're not prepared to\b",
]

# List of words that suggest the answer needs specific financial numbers to properly answer the question.
SPECIFICITY_TRIGGERS = [
    "how much","how many","what is","what are","when","which","where","who","why",
    "range","guidance","margin","capex","opex","revenue","sales","eps","ebitda",
    "timeline","date","target","growth","update","split","dividend","cost","price",
    "units","volumes","gross","net","tax","percentage","utilization","order book"
]

NUMERIC_PATTERN = r"(?:\d+(?:\.\d+)?%|\b\d{1,3}(?:,\d{3})*(?:\.\d+)?\b|£|\$|€)"

In [79]:
# Function to calculate cosine similarity between question and answers.
def cosine_sim(q, a):
    vec = TfidfVectorizer(stop_words='english').fit_transform([q, a]) # converts text to vectors 
    sim = float(cosine_similarity(vec[0], vec[1])[0, 0]) # calculate the cosine similarity between the two vectors

    return sim

In [80]:
# Function to compute baseline evasion score.
def baseline_evasion_score(q, a):
    # 1. Cosine similarity
    sim = cosine_sim(q, a) # calculates cosine similarity using previous function
    sim_component = (1 - sim) * 45 # less similar the answer is, the bigger the contribution to the evasion score, scaled by 45

    # 2. Numerical specificity- Does the question require and answer with financial figures/ a specific answer?
    needs_num = any(t in q.lower() for t in SPECIFICITY_TRIGGERS) # true if the question requires a numeric/ specific answer
    has_num = bool(re.search(NUMERIC_PATTERN, a)) # true if the answer includes a number 
    numeric_component = 25 if needs_num and not has_num else 0 # score of 25 if the question needs a number but the answer doesn't give one

    # 3. Evasive phrases- does the answer contain evasive phrases?
    phrase_hits = sum(len(re.findall(p, a.lower())) for p in EVASIVE_PHRASES) # counts how many times an evasive phrase appears in the answer
    phrase_component = min(3, phrase_hits) * 8 # max of 3 hits counted, each hit = 8 points 

    # Final evasion score.
    score = min(100, sim_component + numeric_component + phrase_component) # adds components together and caps score at 100
    
    return score, sim, phrase_hits, needs_num, has_num

### **LLM and Blended Evasion Score Functions**

In [81]:
# Function to build the premise for the model (question + answer).
def build_premise(q, a):
    return f'[QUESTION] {q} [ANSWER] {a}'

In [93]:
# Function to return probability of entailment, neutral and contradiction.
def llm_probabilities(model, tokenizer, device, premise, hypothesis):
    model.eval()
    encoder = tokenizer(premise, hypothesis, return_tensors='pt', truncation=True, max_length=512)
    encoder = {k: v.to(device) for k, v in encoder.items()}
    with torch.no_grad():
        logits = model(**encoder).logits.squeeze(0)
    raw = F.softmax(logits, dim=-1).tolist()

    # Label order.
    id2label = {0: 'contradiction', 1: 'neutral', 2: 'entailment'}

    # If num_labels = 3
    if len(raw) == 3:
        table = {id2label[i]: raw[i] for i in range(3)}
        return {
            'entailment': table.get('entailment', 0.0),
            'neutral': table.get('neutral', 0.0),
            'contradiction': table.get('contradiction', 0.0)
        }

    # If num_labels = 2 (treat as entailment vs not entailment)
    elif len(raw) == 2:
        table = {id2label[i]: raw[i] for i in range(2)}
        ent = table.get('entailment', raw[1])
        not_ent = table.get('entailment', raw[0])
        return {
            'entailment': ent,
            'neutral': 0.0, 
            'contradiction': not_ent
        }
    else:
        return {'entailment': 0.0,
                'neutral': 1.0, 
                'contradiction': 0.0}

In [94]:
# Function compute llm evasion scores.
def llm_evasion_score(question, answer, model, tokenizer, device):
    premise = build_premise(question, answer)

    # Hypothesis.
    H_DIRECT = 'The answer gives a direct and specific response to the question'
    H_EVASIVE = 'The answer avoids giving a direct response to the question'

    pd = llm_probabilities(model, tokenizer, device, premise, H_DIRECT)
    pe = llm_probabilities(model, tokenizer, device, premise, H_EVASIVE)

    e_direct = max(1e-9, pd['entailment'])
    e_evasive = max(1e-9, pe['entailment'])

    # Prevent both hypothesis being simultaneously high/low.
    denom = math.exp(e_direct) + math.exp(e_evasive)
    p_direct = math.exp(e_direct) / denom
    p_evasive = math.exp(e_evasive) / denom

    return {
        'p_direct': p_direct,
        'p_evasive': p_evasive
    }

In [96]:
# Function to compute blended evasion score and return all scores.
def compute_all_evasion_scores(q, a, *, models_and_tokenizers=models_and_tokenizers, device, LLM_WEIGHT=0.30):
    
    # Compute baseline evasion score.
    base_score, _, _, _, _ = baseline_evasion_score(q, a)

    # Individual LLM scores.
    llm_scores = {}
    for name, (m, t) in models_and_tokenizers.items():
        scores = llm_evasion_score(q, a, m, t, device)
        llm_scores[name] = float(100.0 * scores['p_evasive'])

    # Ensemble LLM score.
    llm_avg = float(np.mean(list(llm_scores.values()))) if llm_scores else 0.0

    # Compute blended score.
    blended_score = float(np.clip((1.0 - LLM_WEIGHT) * base_score + LLM_WEIGHT * llm_avg, 0.0, 100.0))

    return {
        'baseline': base_score,
        'llm_individual': llm_scores,
        'llm_avg': llm_avg,
        'blended': blended_score
        }

### **Main Pipeline**

In [ ]:
# # Define preliminary thresholds.
# LLM_WEIGHT = 0.30
# EVASION_THRESHOLD_BASE = 60.0
# EVASION_THRESHOLD_LLM = 50.0
# EVASION_THRESHOLD_BLENDED = 60.0

In [ ]:
# Function to label 'Direct' or 'Evasive' based on the score.
def label_from_score(score, threshold):
    return 'Evasive' if score >= threshold else 'Direct'

In [99]:
# Evasion Pipeline.
def evasion_pipeline(df, models_and_tokenizers, device, LLM_WEIGHT, EVASION_THRESHOLD_BASE, EVASION_THRESHOLD_LLM, EVASION_THRESHOLD_BLENDED):

    records = []

    for _, row in df.iterrows():
        q, a = str(row['question']), str(row['answer'])
        output = compute_all_evasion_scores(q=q, a=a, LLM_WEIGHT=LLM_WEIGHT, models_and_tokenizers=models_and_tokenizers, device=device)

        pred_base = label_from_score(output['baseline'], EVASION_THRESHOLD_BASE)
        pred_llm_avg = label_from_score(output['llm_avg'], EVASION_THRESHOLD_LLM)
        pred_blended = label_from_score(output['blended'], EVASION_THRESHOLD_BLENDED)

        record = {
            'question_number': row.get('question_number'),
            'question': q,
            'answer': a,

            # Evasion Scores
            'evasion_score_baseline': int(output['baseline']),
            'evasion_score_llm_avg': int(output['llm_avg']),
            "evasion_score_blended": int(output['blended']),

            # Predicted labels.
            'prediction_baseline': pred_base,
            'prediction_llm_avg': pred_llm_avg,
            'prediction_blended': pred_blended,
        }

        for model_name, score in output['llm_individual'].items():
            record[f'evasion_score_{model_name}'] = int(score)
            record[f'prediction_{model_name}'] = label_from_score(score, EVASION_THRESHOLD_LLM)

        records.append(record)

    return pd.DataFrame(records)


### **Fine-Tune Score Thresholds**

### **FORMAT FOR EVASION PIPELINE**
- Prepare data splits (labelled data only): make a validation dataset for threshold search and model selection, test set for final reporting
- Run evasion pipeline on validation set once- save result to disk
- Tune thresholds per detector on the validation data set 
- Determine whether to use ensemble llm or individual llm and get best thresholds
- Apply these locked thresholds and llm model to produce predictions on the test set & evaluate
- Apply these to training set to predict labels.

In [100]:
# Run evasion pipeline.
jpm_2025_evasion_results = evasion_pipeline(
    jpm_2025_qa_pairs_df, 
    models_and_tokenizers=models_and_tokenizers, 
    device=device,
    LLM_WEIGHT=0.30, 
    EVASION_THRESHOLD_BASE=60.0,
    EVASION_THRESHOLD_LLM=50.0,
    EVASION_THRESHOLD_BLENDED=60.0)

# View results.
jpm_2025_evasion_results.head()

question_number                                           question  \
0                1  Good morning, Jeremy. Wondering if you could s...   
1                2  Yeah. And just one question on the NII ex. Mar...   
2                3  Yes. Good morning. This question is for Jamie....   
3                4  Got it. And a second follow-up question. And I...   
4                5  Thank you. Operator: Thank you. Our next quest...   

                                              answer  evasion_score_baseline  \
0  Sure, Ken. So I mean, at a high level, I would...                      64   
1  Yeah, that's a good question, Ken. You're righ...                      44   
2  I just – before Jamie answers that, Erika, I j...                      35   
3  Yeah, Erika, it's a good question. But the tru...                      88   
4  Thanks, Erika. Operator: I apologize. Our next...                      67   

   evasion_score_llm_avg  evasion_score_blended prediction_baseline  \
0                     52                     60             Evasive   
1                     56                     47              Direct   
2                     52                     40              Direct   
3                     49                     76             Evasive   
4                     51                     62             Evasive   

  prediction_llm_avg prediction_blended  evasion_score_roberta  \
0            Evasive            Evasive                     51   
1            Evasive             Direct                     52   
2            Evasive             Direct                     49   
3             Direct            Evasive                     50   
4            Evasive            Evasive                     50   

  prediction_roberta  evasion_score_deberta prediction_deberta  \
0            Evasive                     56            Evasive   
1            Evasive                     54            Evasive   
2             Direct                     61            Evasive   
3            Evasive                     66            Evasive   
4            Evasive                     64            Evasive   

   evasion_score_zs_deberta prediction_zs_deberta  
0                        50               Evasive  
1                        61               Evasive  
2                        48                Direct  
3                        31                Direct  
4                        39                Direct

- Built a validation set of 26 examples from 2023 1q jpm results, human-labelled these evasive or direct.

In [101]:
# Load 2023 transcript.
jpm_1q_23_df = pd.read_csv('../data/processed/jpm/jpm-1q23-earnings-call-transcript_qa.csv')

In [102]:
# View speaker roles.
jpm_1q_23_df['role'].unique()

array(['Chief Financial Officer', 'analyst',
       'Chairman & Chief Executive Officer'], dtype=object)

In [103]:
# Define role mapping.
role_map = {
    'analyst': 'analyst',
    'Chief Financial Officer': 'banker',
    'Chairman & Chief Executive Officer': 'banker'
}

# Apply to dataset.
jpm_1q_23_df['role_normalised'] = jpm_1q_23_df['role'].map(role_map)

# View the dataset.
jpm_1q_23_df.head()

section  question_number  answer_number   speaker_name  \
0  presentation              NaN            NaN  Jeremy Barnum   
1            qa              NaN            NaN  Steven Chubak   
2            qa              NaN            NaN  Jeremy Barnum   
3            qa              1.0            NaN  Steven Chubak   
4            qa              1.0            1.0    Jamie Dimon   

                                 role               company  \
0             Chief Financial Officer         JPMorganChase   
1                             analyst    Wolfe Research LLC   
2             Chief Financial Officer  JPMorgan Chase & Co.   
3                             analyst    Wolfe Research LLC   
4  Chairman & Chief Executive Officer  JPMorgan Chase & Co.   

                                             content  year quarter  \
0  Thanks, and good morning, everyone. The presen...  2023      Q1   
1                                 Hey, good morning.  2023      Q1   
2                               Good morning, Steve.  2023      Q1   
3  So, Jamie, I was actually hoping to get your p...  2023      Q1   
4  Well, I think you were already kind of complet...  2023      Q1   

   is_pleasantry role_normalised  
0          False          banker  
1           True         analyst  
2           True          banker  
3          False         analyst  
4          False          banker

In [104]:
# Filter out presention and is_pleasantry == True
jpm_1q_23_df = jpm_1q_23_df[jpm_1q_23_df['section'] == 'qa']
jpm_1q_23_df = jpm_1q_23_df[jpm_1q_23_df['is_pleasantry'] == False]

# View the dataset.
jpm_1q_23_df

section  question_number  answer_number   speaker_name  \
3       qa              1.0            NaN  Steven Chubak   
4       qa              1.0            1.0    Jamie Dimon   
5       qa              1.0            1.0  Steven Chubak   
6       qa              1.0            2.0    Jamie Dimon   
7       qa              1.0            2.0  Steven Chubak   
..     ...              ...            ...            ...   
93      qa             26.0            NaN  Matt O'Connor   
94      qa             26.0            1.0    Jamie Dimon   
95      qa             26.0            2.0  Jeremy Barnum   
96      qa             26.0            3.0    Jamie Dimon   
97      qa             27.0            NaN  Matt O'Connor   

                                  role                         company  \
3                              analyst              Wolfe Research LLC   
4   Chairman & Chief Executive Officer            JPMorgan Chase & Co.   
5                              analyst              Wolfe Research LLC   
6   Chairman & Chief Executive Officer            JPMorgan Chase & Co.   
7                              analyst              Wolfe Research LLC   
..                                 ...                             ...   
93                             analyst  Deutsche Bank Securities, Inc.   
94  Chairman & Chief Executive Officer            JPMorgan Chase & Co.   
95             Chief Financial Officer            JPMorgan Chase & Co.   
96  Chairman & Chief Executive Officer            JPMorgan Chase & Co.   
97                             analyst  Deutsche Bank Securities, Inc.   

                                              content  year quarter  \
3   So, Jamie, I was actually hoping to get your p...  2023      Q1   
4   Well, I think you were already kind of complet...  2023      Q1   
5   Got it. And just in terms of appetite for the ...  2023      Q1   
6                                           Oh, yeah.  2023      Q1   
7                    ...elevated macro uncertainties.  2023      Q1   
..                                                ...   ...     ...   
93  Okay. And then just separately to squeeze in –...  2023      Q1   
94  That'll be every quarter for the rest of our l...  2023      Q1   
95                                             Cheap.  2023      Q1   
96                                             Cheap.  2023      Q1   
97  Okay. All right. Thank you. Operator: We have ...  2023      Q1   

    is_pleasantry role_normalised  
3           False         analyst  
4           False          banker  
5           False         analyst  
6           False          banker  
7           False         analyst  
..            ...             ...  
93          False         analyst  
94          False          banker  
95          False          banker  
96          False          banker  
97          False         analyst  

[85 rows x 11 columns]

In [105]:
# Create qa pairs.
jpm_1q_23_qa_pairs_df = create_qa_pairs(jpm_1q_23_df)

# View dataset.
display(jpm_1q_23_qa_pairs_df.head())

# View shape.
print('Number of samples:', jpm_1q_23_qa_pairs_df.shape[0])

question_number                                           question  \
0              1.0  So, Jamie, I was actually hoping to get your p...   
1              2.0  Hey, thanks. Good morning. Hey, Jeremy, I was ...   
2              3.0  Yeah, and as a follow-up on the point about ra...   
3              4.0  Hi, thanks. Jeremy, wanted to follow up again ...   
4              5.0  Okay. And then I wanted to ask Jamie – there's...   

                                              answer  
0  Well, I think you were already kind of complet...  
1  Yeah, sure. So let me just summarize the drive...  
2  Well first of all, I don't quite believe it. S...  
3  Yeah. John, it's a really good question, and w...  
4  Yeah. I wouldn't use the word credit crunch if...

Number of samples: 26


In [106]:
# Create the test set. 
jpm_1q_23_test_set_df = jpm_1q_23_qa_pairs_df.copy()

# Create a blank label column and export to CSV for human to label.
jpm_1q_23_test_set_df['label'] = ''  # fill with Direct or Evasive
jpm_1q_23_test_set_df.to_csv('jpm_1q_23_test_set.csv', index=False)

- The test set was human labelled with either 'direct' or 'evasive'.

In [107]:
# Import the labelled test set. 
jpm_1q_23_test_set_labelled_df = pd.read_csv('../notebooks/summarisation_evasion_files/jpm_1q_23_test_set_labelled.csv')

# View dataset.
jpm_1q_23_test_set_labelled_df.head()

question_number                                           question  \
0                1  So, Jamie, I was actually hoping to get your p...   
1                2  Hey, thanks. Good morning. Hey, Jeremy, I was ...   
2                3  Yeah, and as a follow-up on the point about ra...   
3                4  Hi, thanks. Jeremy, wanted to follow up again ...   
4                5  Okay. And then I wanted to ask Jamie – there's...   

                                              answer    label  
0  Well, I think you were already kind of complet...  Evasive  
1  Yeah, sure. So let me just summarize the drive...   Direct  
2  Well first of all, I don't quite believe it. S...   Direct  
3  Yeah. John, it's a really good question, and w...  Evasive  
4  Yeah. I wouldn't use the word credit crunch if...   Direct

In [108]:
# Run evasion pipeline.
jpm_1q_23_evasion_results = evasion_pipeline(jpm_1q_23_test_set_labelled_df)

# Reappend the human label.
jpm_1q_23_evasion_results['human_label'] = jpm_1q_23_test_set_labelled_df['label']

# View results.
jpm_1q_23_evasion_results.head()

TypeError: evasion_pipeline() missing 6 required positional arguments: 'models_and_tokenizers', 'device', 'LLM_WEIGHT', 'EVASION_THRESHOLD_BASE', 'EVASION_THRESHOLD_LLM', and 'EVASION_THRESHOLD_BLENDED'

In [ ]:
# Function to evaluate the evasion scores vs true labels.
def evaluate_evasion_scores(df):

    # True labels: 1 = Evasive, 0 = Direct (using 'human_label').
    y_true = (df["human_label"].astype(str).str.strip().str.lower() == 'evasive').astype(int).values

    # Convert predicted label strings to binary (1 = Evasive, 0 = Direct).
    def to_binary(pred_series):
        return (pred_series.astype(str).str.strip().str.lower() == 'evasive').astype(int).values

    # Convert predicted labels to binary.
    y_pred_base  = to_binary(df["prediction_baseline"])
    y_pred_llm   = to_binary(df["prediction_llm"])
    y_pred_blend = to_binary(df["prediction_blended"])

    return {
        'baseline': {
            'classification_report': classification_report(y_true, y_pred_base, target_names=["Direct", "Evasive"], digits=3, zero_division=0),
            'confusion_matrix': confusion_matrix(y_true, y_pred_base)
        },
        'llm': {
            'classification_report': classification_report(y_true, y_pred_llm, target_names=["Direct", "Evasive"], digits=3, zero_division=0),
            'confusion_matrix': confusion_matrix(y_true, y_pred_llm)
        },
        'blended': {
            'classification_report': classification_report(y_true, y_pred_blend, target_names=["Direct", "Evasive"], digits=3, zero_division=0),
            'confusion_matrix': confusion_matrix(y_true, y_pred_blend) 
        }
    }

In [ ]:
# Extract results.
eval_dict = evaluate_evasion_scores(jpm_1q_23_evasion_results)
baseline_eval, llm_eval, blended_eval = eval_dict['baseline'], eval_dict['llm'], eval_dict['blended']

In [ ]:
# View baseline results.
base_cr, base_cm = baseline_eval['classification_report'], baseline_eval['confusion_matrix']

print(base_cr)
print(base_cm)

              precision    recall  f1-score   support

      Direct      0.588     0.714     0.645        14
     Evasive      0.556     0.417     0.476        12

    accuracy                          0.577        26
   macro avg      0.572     0.565     0.561        26
weighted avg      0.573     0.577     0.567        26

[[10  4]
 [ 7  5]]


In [ ]:
# View llm results.
llm_cr, llm_cm = llm_eval['classification_report'], llm_eval['confusion_matrix']

print(llm_cr)
print(llm_cm)

              precision    recall  f1-score   support

      Direct      0.000     0.000     0.000        14
     Evasive      0.462     1.000     0.632        12

    accuracy                          0.462        26
   macro avg      0.231     0.500     0.316        26
weighted avg      0.213     0.462     0.291        26

[[ 0 14]
 [ 0 12]]


In [ ]:
# View blended results.
blended_cr, blended_cm = blended_eval['classification_report'], blended_eval['confusion_matrix']

print(blended_cr)
print(blended_cm)

              precision    recall  f1-score   support

      Direct      1.000     0.071     0.133        14
     Evasive      0.480     1.000     0.649        12

    accuracy                          0.500        26
   macro avg      0.740     0.536     0.391        26
weighted avg      0.760     0.500     0.371        26

[[ 1 13]
 [ 0 12]]


- Baseline (rule-based) accuracy is currently the highest at 60%
- Fine tune the thresholds using grid search to determine the optimal threshold to give the best accuracy.

### **Tuning threshold**

- Recall will be the priority over accuracy to prevent missing evasive answers. 
- Some false positives are tolerable and regulator can review flagged answers. 

In [ ]:
# Obtain validation set with evasion scores and drop predictions.
jpm_1q_23_tuning = jpm_1q_23_evasion_results.drop(['prediction_baseline', 'prediction_llm', 'prediction_blended'], axis=1)
jpm_1q_23_tuning.head()

question_number                                           question  \
0                1  So, Jamie, I was actually hoping to get your p...   
1                2  Hey, thanks. Good morning. Hey, Jeremy, I was ...   
2                3  Yeah, and as a follow-up on the point about ra...   
3                4  Hi, thanks. Jeremy, wanted to follow up again ...   
4                5  Okay. And then I wanted to ask Jamie – there's...   

                                              answer  evasion_score_baseline  \
0  Well, I think you were already kind of complet...                      55   
1  Yeah, sure. So let me just summarize the drive...                      80   
2  Well first of all, I don't quite believe it. S...                      40   
3  Yeah. John, it's a really good question, and w...                      78   
4  Yeah. I wouldn't use the word credit crunch if...                      55   

   evasion_score_llm  evasion_score_blended human_label  
0                 79                     79     Evasive  
1                 78                    100      Direct  
2                 83                     65      Direct  
3                 73                     99     Evasive  
4                 81                     80      Direct

In [ ]:
# Function to extract ground truth (1 = Evasive, 0 = Direct)
def extract_y_true(df):
    return (df['human_label'].astype(str).str.strip().str.lower() == 'evasive').astype(int).values

In [ ]:
# Function calculate metrics for each threshold.
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

def tune_threshold(df, score_col, thr_grid):
    y_true = extract_y_true(df)                     # get true labels
    scores = df[score_col].astype(float).values     # get raw evasion scores 

    rows = []
    for thr in thr_grid:
        y_pred = (scores >= thr).astype(int) # label response evasive (1) if score is higher than threshold

        precision = precision_score(y_true, y_pred, zero_division=0)
        recall = recall_score(y_true, y_pred, zero_division=0)
        f1 = f1_score(y_true, y_pred, zero_division=0)
        accuracy = accuracy_score(y_true, y_pred)

        rows.append({
            'threshold': float(thr),
            'precision': precision,
            'recall': recall,
            'f1': f1,
            'accuracy': accuracy
        })
    
    results = pd.DataFrame(rows).sort_values(
        by=['recall', 'f1', 'precision'],
        ascending=[False, False, False]
        ).reset_index(drop=True)
    
    return results

In [ ]:
# Define threshold ranges around current thresholds.
thr_base_grid = np.arange(40, 85, 5)
thr_llm_grid = np.arange(35, 85, 5)
thr_blend_grid = np.arange(40, 85, 5)

In [ ]:
# Run grid search for each detector.
base_results = tune_threshold(jpm_1q_23_tuning, 'evasion_score_baseline', thr_base_grid)
llm_results = tune_threshold(jpm_1q_23_tuning, 'evasion_score_llm', thr_llm_grid)
blend_results = tune_threshold(jpm_1q_23_tuning, 'evasion_score_blended', thr_blend_grid)

In [ ]:
# Extract the best thresholds.
best_base_thr = base_results.loc[0, 'threshold']
best_llm_thr = llm_results.loc[0, 'threshold']
best_blend_thr = blend_results.loc[0, 'threshold']

print('Best Baseline Threshold:', best_base_thr)
print('Best LLM Threshold:', best_llm_thr)
print('Best Blended Threshold', best_base_thr)

Best Baseline Threshold: 40.0
Best LLM Threshold: 70.0
Best Blended Threshold 40.0


In [ ]:
# Inspect trade-offs.
print('\nTop 5 baseline configs:\n', base_results.head())
print('\nTop 5 llm configs:\n', llm_results.head())
print('\nTop 5 blended configs:\n', blend_results.head())


Top 5 baseline configs:
    threshold  precision    recall        f1  accuracy
0       40.0   0.454545  0.833333  0.588235  0.461538
1       45.0   0.500000  0.666667  0.571429  0.538462
2       50.0   0.583333  0.583333  0.583333  0.615385
3       55.0   0.545455  0.500000  0.521739  0.576923
4       65.0   0.714286  0.416667  0.526316  0.653846

Top 5 llm configs:
    threshold  precision  recall        f1  accuracy
0       70.0   0.521739     1.0  0.685714  0.576923
1       65.0   0.480000     1.0  0.648649  0.500000
2       35.0   0.461538     1.0  0.631579  0.461538
3       40.0   0.461538     1.0  0.631579  0.461538
4       45.0   0.461538     1.0  0.631579  0.461538

Top 5 blended configs:
    threshold  precision  recall        f1  accuracy
0       55.0   0.480000     1.0  0.648649  0.500000
1       60.0   0.480000     1.0  0.648649  0.500000
2       40.0   0.461538     1.0  0.631579  0.461538
3       45.0   0.461538     1.0  0.631579  0.461538
4       50.0   0.461538     1.0 

**Baseline:**
- 40.0 gave the highest recall (0.83) but lower precision (0.45)
- 50.0 is more balanced (0.58 recall and 0.58 precision)

**LLM**
- Every top configuration has recall = 1.0 which shows the LLM detector is very sensitive 
- Raising the threshold increases precision slightly = 70.0

**Blended**
- Similar to LLM result, 55.0 or 60.0 gave slightly higher precision